FreeRTOS/FreeRTOS-Kernel
# Detect more startup config errors on Cortex M

Description
-----------
Add new assert statements in `xPortStartScheduler()` on Cortex M to verify correct installation of the PendSV and SVCall handlers.  These same assertion statements also verify VTOR is set properly for targets with a modifiable VTOR register.  These are common causes of startup issues for Cortex M users.

On ARMv7-M and ARMv8-M, add explicit setting of SVCall priority to 0 (the maximum priority).  Prior to this PR the port code depended on system startup code not to change the priority from its default of 0.  This a rare cause of startup issues for Cortex M users. (See last link below.)  For consistency, replicate this new method of setting SVCall priority to the ARMv7-M MPU ports, replacing the previous mechanism used by those ports to set SVCall priority.

Test Steps
-----------
ARMv7-M: Execute demo application with and without `#define vPortSVCHandler SVC_Handler` and `#define xPortPendSVHandler PendSV_Handler` in FreeRTOSConfig.h.  Before this PR, the application works correctly with both #defines, and doesn't work correctly without either one.  After this PR, the application still works correctly with both #defines, and has an assert failure without either one, as intended.

ARMv6-M: Same as above but the port doesn't use vPortSVCHandler; test with and without xPortPendSvHandler defined.  XMC1000 demo cannot be used for this test.

ARMv7-M / ARMv8-M: Use demo application.  First thing in main() set VTOR to the wrong value.  Before this PR, xPortStartScheduler() hard faults.  After this PR, an assert failure detects the issue, as intended.

Checklist:
----------

Related Issues
-----------
https://forums.freertos.org/t/prvportstartfirsttask-fails-at-svc-0/18288
https://forums.freertos.org/t/hard-fault-encountered-after-vtaskstartscheduler-is-invoked/17476
https://forums.freertos.org/t/freertos-does-not-execute-first-task-after-svc-0-call/16124
https://forums.freertos.org/t/xportstartscheduler-fails-because-basepri-is-non-zero/15509
...

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
